fem/bilininteg_hcurl.cpp

```diff
@@ -1908,4 +1928,1284 @@ void PAHcurlH1Apply2D(const int D1D,
    }); // end of element loop
 }
 
+// PA H(curl) Mass Assemble 3D kernel
+static void PAHcurlL2Setup3D(const int Q1D,
+                             const int coeffDim,
+                             const int NE,
+                             const Array<double> &w,
+                             Vector &_coeff,
+                             Vector &op)
+{
+   const int NQ = Q1D*Q1D*Q1D;
+   auto W = w.Read();
+   auto coeff = Reshape(_coeff.Read(), coeffDim, NQ, NE);
+   auto y = Reshape(op.Write(), coeffDim, NQ, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      for (int q = 0; q < NQ; ++q)
+      {
+         for (int c=0; c<coeffDim; ++c)
+         {
+            y(c,q,e) = W[q] * coeff(c,q,e);
```

Comment:
In the code I work on we have seen improved performance by flattening the iteration space then converting from 1D to 3D indexing. 
```
MFEM_FORALL(id,NE*NQ*coeffDim,
{... 
}) 
```

Reply:
That is an interesting point. Is it faster because it gives the GPU more small tasks? I think most of the `MFEM_FORALL` loops in PA kernels are just over the elements (`NE`), so this comment could apply to a lot of code. I am not sure whether this PR should address this, but maybe we should do some experiments with various kernels to see if performance can be improved, and if so, make the changes in another PR.

Reply:
It definitely exposes more parallelism and should help. I think making all the changes is outside of the scope of this PR.